[team-4 롤로] 장소 검색 화면 & 네트워크 모델(+ Fake Data)

eeeesong opened the pull request:
gangwoon 님 안녕하세요, iOS 클래스 롤로라고 합니다 :)
장소 검색 화면까지 구현한 결과물에 대해 PR을 요청드리려고 합니다.

----

## 현재 화면

![0521 airbnb](https://user-images.githubusercontent.com/72188416/119091389-f45b4a80-ba47-11eb-8af1-e9f345d597d0.gif)


## 구현 내용

### Home 화면
- [x] navigation controller 적용
- [x] tabBar controller 적용 
- [x] 히어로 이미지 영역 구성
- [x] navigation에 searchBar 추가
- [x] 검색 바 영역 터치 시 검색 화면으로 전환  

### Location Search 화면
- [x] navigation에 search controller 적용
- [x] 인기 여행지 table view 구성
- [x] 검색 화면 진입 시 키보드 노출
- [x] 검색 키워드 입력 시 result table view로 전환 (UISearchResultsUpdating 사용)
- [x] 검색 시작 시 `지우기` 버튼 활성화
- [x] 검색 결과 터치 시 캘린더 선택 화면으로 전환 

### Network Model 구축
- [x] Alamofire 사용
- [x] RequestManager, NetworkManager, ImageLoadManager 생성
- [x] CustomError 생성, 에러 발생 시 Alert로 문제 상황 알림 
- [x] Fake Network 객체를 통해 가짜 데이터를 화면에 표시 (API 미구현 상태)
- [x] 디바이스에 이미지 데이터를 캐싱하도록 함


## 고민사항

### ViewModel의 쓰임새
- 지난 미션까지는 ViewModel과 함께 Combine을 사용했고, ViewModel에 Observable한 변수들을 저장하여 ViewController가 값을 Observe하여 반영하도록 구현을 하였습니다. 그런데 이번 미션에선 의도적으로 Combine을 쓰지 않고 모두 탈출 클로저로 처리를 하였는데, 그러다 보니 ViewModel이 메소드를 통해 값들을 전달하게 되어 ViewModel 자체의 역할이 애매모호해진 느낌이 있습니다. 🥲 
- 지금의 상태라면 ViewModel이나 UseCase 둘 중 하나가 없어지고 오히려 Repository 역할을 하는 객체가 있는 게 맞겠다는 생각이 들기도 하지만, 앞으로의 구현 사항을 고려하여 우선 지금의 상태를 유지해두었습니다. 

 ----


앞으로 3주 동안 잘 부탁드립니다!
감사합니다 :)

Dae-Hwa committed:
chore: .gitignore 생성

Dae-Hwa committed:
chore: 파트별 폴더 생성

Dae-Hwa committed:
chore: notion 링크 추가

eeeesong committed:
chore : 프로젝트 파일 생성

eeeesong committed:
feat: tab bar 생성

- 검색, 위시리스트, 로그인에 해당하는 심볼 설정

Dae-Hwa/airbnb/#1

eeeesong committed:
feat: 콘텐츠 영역 구성

- Search bar 추가
- 히어로 이미지 영역 추가

Dae/Hwa/airbnb/#1

eeeesong committed:
feat: 오토레이아웃 적용

- 모든 아이폰 화면 대응

Dae-Hwa/airbnb/#1

eeeesong committed:
Merge pull request #5 from Dae-Hwa/iOS/feature/main-layout

[iOS] 메인 UI 디자인

eeeesong committed:
chore: 디렉토리 변경

eeeesong committed:
feat: 검색 화면 구성

- 서치 바 추가
- 인기 여행지 테이블 추가

Dae-Hwa/airbnb/#6

eeeesong committed:
feat: 검색바 터치 시 화면 전환

- 네비게이션 컨트롤러 추가
- 검색 바 터치 시 PopularLocationViewController로 전환

Dae-Hwa/airbnb/#6

eeeesong committed:
feat: 근처의 인기 여행지 리스트 화면 구성

- 헤더, 셀 추가
- 더미 데이터 구성

Dae-Hwa/airbnb/#6

Dae-Hwa committed:
chore: 커밋 템플릿 추가

eeeesong committed:
feat: 검색 바 UI 변경

- UISearchController 적용

Dae-Hwa/airbnb/#6

eeeesong committed:
feat: 검색 결과 화면 구성

- 검색 시작 시 검색 결과 테이블 뷰 노출

Dae-Hwa/airbnb/#6

eeeesong committed:
feat: cancel 버튼으로 검색 초기화

- 검색 시작 시 cancel 버튼 노출
- search controller inactive하여 memory leak 되지 않도록 함

Dae-Hwa/airbnb/#6

eeeesong committed:
fix: 네비게이션 트랜지션 버그 해결

- 배경 색상을 흰색으로 변경
- extend edge 설정 변경

Dae-Hwa/airbnb/#15

eeeesong committed:
fix: return 시 Cancel 버튼 소멸 버그 수정

Dae-Hwa/airbnb/#6

Dae-Hwa committed:
chore: README.md에 커밋 템플릿 관련 내용 추가

Dae-Hwa committed:
docs: README에 팀원 및 팀 문서 링크 추가

Dae-Hwa committed:
chore: 커밋 템플릿에 타입 설명 추가

Dae-Hwa committed:
docs: README.md에 커밋 템플릿 예시 사진 업데이트

eeeesong committed:
feat: 검색 결과 선택 시 달력 화면 이동

Dae-Hwa/airbnb/#6

eeeesong committed:
chore: 파일 디렉토리 변경

eeeesong committed:
Merge branch 'team-4' of https://github.com/Dae-Hwa/airbnb into iOS/dev

eeeesong committed:
Merge branch 'iOS/dev' of https://github.com/Dae-Hwa/airbnb into iOS/feature/search-screen

eeeesong committed:
fix: 브랜치 충돌 해결

- squash로 인한 충돌

eeeesong committed:
Merge pull request #16 from Dae-Hwa/iOS/feature/search-screen

[iOS] 지역 검색 화면 디자인

eeeesong committed:
chore: Alamofire 패키지 추가

- Swift Package Manager Dependency

Dae-Hwa/airbnb/#17

eeeesong committed:
feat: Alamofire를 통한 네트워크 모델 생성

- Request 생성, Network 요청 전송, ImageDownload 모델 생성

Dae-Hwa/airbnb/#17

eeeesong committed:
feat: 인기 여행지 Mock Data 연동

- ViewModel, UseCase 흐름 구축

Dae-Hwa/airbnb/#17

eeeesong committed:
chore: Mock data 내용 변경

Dae-Hwa/airbnb/#17

eeeesong committed:
feat: 위치 검색 Mock Data 연동

- ViewModel, UseCase 흐름 생성

Dae-Hwa/airbnb/#17

eeeesong committed:
refactor: view controller 생성 방식 개선

- 팩토리 메소드를 통해 생성

Dae-Hwa/airbnb/#17

eeeesong committed:
refactor: 프로토콜 생성

- 의존성 개선

Dae-Hwa/airbnb/#17

eeeesong committed:
feat: 에러 처리 개선

- get 실패 시 alert 띄우기

Dae-Hwa/airbnb/#17

eeeesong committed:
feat: 가짜 네트워크 객체 생성

- 200 코드와 가짜 데이터를 받아와 화면에 표시한다.

Dae-Hwa/airbnb/#17

eeeesong committed:
feat: 이미지 다운로드, 캐싱 구현

- Hero Image와 Popular Location 이미지를 받아올 수 있도록 한다.
- 이미지 url은 임시 사용

Dae-Hwa/airbnb/#17

eeeesong committed:
chore: final, private 키워드 추가

eeeesong committed:
Merge pull request #20 from Dae-Hwa/iOS/feature/network-model

[iOS] 네트워크 모델 구현 + MVVM/UseCase 적용
Closes #17

eeeesong committed:
fix: 강한 참조로 인한 memory leak 해결

- 클로저에서 사용되는 self를 weak하게 변경
- delegate를 weak var로 정의

Dae-Hwa/airbnb/#21

eeeesong committed:
Merge pull request #22 from Dae-Hwa/iOS/fix/memory-leak

[iOS] fix: 강한 참조로 인한 memory leak 해결

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/CalendarSelect/CalendarViewController.swift:
```diff
@@ -0,0 +1,21 @@
+final class CalendarViewController: UIViewController, Instantiable {
+
+    static var reuseIdentifier: String { String(describing: self) }
+    private let backButtonTitle = "날짜 선택"
```
VIEW에서 갖고 있는 것 보다 viewModel에서 데이터를 갖고 있는게 좋을 거 같습니다.
뷰는 오직 데이터를 받아서 화면에 보여주는 역활을 했으면 좋겠습니다.

eeeesong replied:
해당 데이터를 viewModel로 옮겼습니다!

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/Entity/LocationSearchResult.swift:
```diff
@@ -0,0 +1,23 @@
+    let latitude: Double
+    let longitude: Double
+    
+    init() {
```
꼭 필요한 생성자일까요?

eeeesong replied:
fake 데이터를 수월하게 생성하기 위해 넣어두었는데, 추후 필요하지 않을 것 같아서 삭제했습니다!

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/View/LocationSearchBar.swift:
```diff
@@ -0,0 +1,29 @@
+
+import UIKit
+
+final class LocationSearchBar: UISearchBar {
```
UISearchBar를 상속받지 않고도 해당 작업을 처리할 수 있을 거 같습니다.

eeeesong replied:
UISearchBar 상속 없이 처리하도록 개선하였습니다 :)

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/View/PopularLocationTableViewCell.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
```
작성한 이유가 불분명한 코드는 제거해주시기 바랍니다.

eeeesong replied:
제거하였습니다!

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/PopularLocationTableViewDataSource.swift:
```diff
@@ -0,0 +1,45 @@
+        let cellID = PopularLocationTableViewCell.reuseIdentifier
+        let cell = tableView.dequeueReusableCell(withIdentifier: cellID) as? PopularLocationTableViewCell ?? PopularLocationTableViewCell()
+        
+        guard let popularLocations = popularLocations else { return cell }
```
제 개인적인 생각입니다.

가드문이 여러 번 존재하는데, 데이터를 온전히 받아오지 못하면 애당초에 해당 셀을 보여주지 않는게 맞다고 생각하는데 롤로는 어떠신가요?

eeeesong replied:
텍스트 정보, 이미지 정보가 있는데 만약 이미지를 받아올 수 없는 상황이 됐을 때
텍스트만이라도 표시해주는 게 낫다고 판단을 했습니다!

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/PopularLocationTableViewDataSource.swift:
```diff
@@ -0,0 +1,45 @@
+        guard let popularLocations = popularLocations else { return cell }
+        
+        let targetIndex = indexPath.row
+        let textInfos = popularLocations[targetIndex]
```
이름이 너무 제네릭합니다.
조금 더 구체적인 이름을 갖게되면 좋을 거 같습니다.

eeeesong replied:
구체적인 이름으로 수정하였습니다.

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/PopularLocationViewController.swift:
```diff
@@ -0,0 +1,154 @@
+    private let backButtonTitle = "위치 검색"
+    
+    @IBOutlet weak var popularLocationTableView: UITableView!
+    private var popularLocationTableViewDataSource: PopularLocationTableViewDataSource!
```
강제 언래핑을 사용하신 이유는 어느정도 유추할 수 있지만, 문제가 생겼을 때 그 이유만으로 상대방을 납득시킬 순 없으실 겁니다.
되도록 강제 언래핑은 피하셨으면 좋겠습니다.

eeeesong replied:
강제 언래핑을 사용하지 않도록 수정했습니다!

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/PopularLocationViewController.swift:
```diff
@@ -0,0 +1,154 @@
+    
+    @IBOutlet weak var popularLocationTableView: UITableView!
+    private var popularLocationTableViewDataSource: PopularLocationTableViewDataSource!
+    private var searchController: LocationSearchController!
```
해당 객체들을 뷰모델로 이동시키는 건 어떨까요?

eeeesong replied:
DataSource들을 뷰모델로 옮기는 대신, 뷰모델에서 데이터를 전부 관리하고 변경이 있을 때마다
뷰컨트롤러가 Observe할 수 있게 하고, 그 때마다 DataSource의 데이터들도 업데이트할 수 있도록 수정하였습니다!

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/PopularLocationViewController.swift:
```diff
@@ -0,0 +1,154 @@
+    }
+    
+    private func configurePopularLocationTableView() {
+        viewModel.popularLocations { [weak self] result in
```
위에서 언급하셨던 viewModel이 어색하다고 느끼신 부분에 시작점입니다.
이 부분을 추상화 시키시면 viewModel이 애매한 포지션이 아니게 될겁니다.

eeeesong replied:
넵! viewModel의 역할을 개선해보도록 하겠습니다

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/PopularLocationViewController.swift:
```diff
@@ -0,0 +1,154 @@
+    }
+    
+    private func startDownloadingImages(from imagePaths: [String]) {
+        var cachePaths = Array(repeating: "", count: imagePaths.count)
```
cachePaths를 미리 만들어 두신 이유가 있을까요?

eeeesong replied:
받아와야하는 정보 개수만큼 배열을 구성해놓고, 이미지 데이터가 받아지는대로 정보를 채워주기 위해 이렇게 작성했습니다

GangWoon replied:
빈 배열을 생성하고 추가해주는 방법은 어떠신가요?
빈 객체를 생성하고 이후에 의미가 있는 데이터를 집어넣는 방식은 위험할 수 있다는게 제 생각입니다. 


예전 objective c 코드를 보면 빈 객체를 생성하고 프로퍼티를 추후에 할당하는 방식을 사용했는데 이 방식은 해당 api 사용 방법을 모르는 사람이 사용했을때 앱크래시가 발생할 수 있습니다. 
해당 코드는 그런 위험이 없는 코드지만 init() 초기화를 만드시는 걸 봤어서 경고차원으로 드렸던 피드백입니다. 꼭 고치시라는건 아니여서 선택하시면 좋을듯 합니다.

eeeesong replied:
이미지 패스를 위한 배열을 따로 생성하지 않고, 기존에 있던 PopularLocation에 패스를 업데이트하여 보여주는 방식으로 바꾸었습니다.
```swift
struct PopularLocation: Decodable {
    let name: String
    let distanceFromHere: String
    let imagePath: String
    var cachePath: String? //추가
}
```

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/PopularLocationViewController.swift:
```diff
@@ -0,0 +1,154 @@
+    private func startDownloadingImages(from imagePaths: [String]) {
+        var cachePaths = Array(repeating: "", count: imagePaths.count)
+        imagePaths.enumerated().forEach { (index, imagePath) in
+            self.viewModel.popularLocationImage(from: imagePath) { [weak self] cachePath in
```
View가 데이터를 받아오는 형태가 아닌 viewModel 데이터르 view에게 데이터를 주입하는 형태가 MVVM에 더 가까울 거 같습니다.
View는 오직 데이터를 보여주는 역활만 해야합니다. ViewModel에서 데이터 변경, 패치 등 여러가지 이벤트가 발생했으면 좋겠습니다.

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/PopularLocationViewController.swift:
```diff
@@ -0,0 +1,154 @@
+    private func pushNextViewController(with result: LocationSearchResult) {
+        let storyboard = self.storyboard ?? StoryboardFactory.create(.accomodationConditions)
+        let nextViewController = ViewControllerFactory.create(from: storyboard, type: CalendarViewController.self)
+        nextViewController.location = result
```
이런 부분들이 강제 언래핑을 제거할 수 있는 기회가 됩니다.

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/SearchResultTableViewController.swift:
```diff
@@ -0,0 +1,71 @@
+import UIKit
+
+protocol SearchResultDelegate: AnyObject {
+    func didSelect(result: LocationSearchResult)
```
위와 같은 이름이 able을 붙인 이름보단 의도를 명확하게 알 수 있을 거 같습니다.
그렇다고 모든 protocol에 delegate를 붙이시란 말은 아닙니다.

eeeesong replied:
-able보다 각 역할을 잘 나타낼 수 있는 네이밍으로 수정하였습니다 :)

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/SearchResultTableViewController.swift:
```diff
@@ -0,0 +1,71 @@
+                let searchResults = try result.get()
+                self?.updateTableView(with: searchResults)
+            } catch {
+                self?.alertError(error: error)
```
이런 부분을 추상화 시키다 보면 viewController가 필요하실 겁니다. 그럴 땐 view마저 추상화 시키시면 좋을 거 같습니다.

eeeesong replied:
View 추상화 부분에 대해서는 이번 수정에서 잘 해결하지 못한 것 같습니다 :(
사실 아직 잘 감이 오지 않아서 좀 더 공부가 필요할 것 같습니다.

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/SearchResultTableViewController.swift:
```diff
@@ -0,0 +1,71 @@
+    }
+    
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return searchResults?.count ?? 0
```
타입이 []?일 필요가 있을까요?

eeeesong replied:
빈 배열로 시작하도록 수정하였습니다!

GangWoon commented on IOS/airbnb/airbnb/AccomodationConditions/LocationSearch/ViewController/SearchResultTableViewController.swift:
```diff
@@ -0,0 +1,71 @@
+    
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cellID = SearchResultTableViewCell.reuseIdentifier
+        let cell = tableView.dequeueReusableCell(withIdentifier: cellID) as? SearchResultTableViewCell ?? SearchResultTableViewCell()
```
코드가 일관성있었으면 좋겠습니다. 어떤 객체는 DataSource가 존재하고 해당 객체는 존재하지 않고 있는 모습입니다.
모든 코드가 동일한 패턴이기 어렵지만 혼자 진행하는 프로젝트에서 상반된 코드가 존재하게 되면 읽는 사람이 의아하게 생각할 수 있을 거 같습니다.

eeeesong replied:
tableViewController로 생성하였기 때문에 따로 DataSource를 두지 않았는데,
말씀해주신대로 일관성 있는 구현이 좋을 것 같아서 따로 생성하는 것으로 수정하였습니다!

GangWoon commented on IOS/airbnb/airbnb/Main/UseCase/TurnOnUseCase.swift:
```diff
@@ -0,0 +1,56 @@
+        self.init(networkManager: FakeNetworkManager(fakeData: fakeData), imageLoadManager: imageLoadManager)
+    }
+    
+    func loadHeroImage(completionHandler: @escaping (Result<String, CustomError>) -> Void) {
```
유사한 일을 하는 UseCase가 다량으로 존재하는 거 같습니다.

코드를 작성하실 때 유사한 코드가 여러곳에 존재하게 되면 코드를 관리하는 입장에서는 코드 품질이 안좋아지고 있다는 적신호를 뜻합니다.

GangWoon commented on IOS/airbnb/airbnb/Main/UseCase/TurnOnUseCase.swift:
```diff
@@ -0,0 +1,56 @@
+    
+    func loadHeroImage(completionHandler: @escaping (Result<String, CustomError>) -> Void) {
+        networkManager.get(decodingType: String.self, endPoint: EndPoint.heroImage) { [weak self] dataResponse in
+            guard let statusCode = dataResponse.response?.statusCode else {
```
해당 가드문을 사용하신 이유를 알 수 있을까요?

eeeesong replied:
넵 인터넷이 연결되지 않았을 경우엔 statusCode 자체가 오지 못할 것 같아서
빠른 종료를 넣고 인터넷 연결 오류를 보여주도록 했습니다

GangWoon commented on IOS/airbnb/airbnb/Main/ViewController/HomeViewController.swift:
```diff
@@ -0,0 +1,68 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        viewModel = HomeViewModel()
+        appDidTurnOn()
```
AppDidTurnOn이름은 AppDelegate에서 호출되는 메소드 처럼 보입니다.

GangWoon commented on IOS/airbnb/airbnb/Network/AlamofireImageLoadManager.swift:
```diff
@@ -0,0 +1,43 @@
+        }
+        
+        let request = downloadRequest(of: imageUrl, fileName: fileName)
+        request.responseURL { response in
```
에러 처리를 UseCase로 갖고 가는것보다 해당 객체에서 핸드링하는게 더 적합해 보입니다.

eeeesong replied:
해당 객체라면 네트워크 객체를 말씀하시는 걸까요??

GangWoon replied:
넵

eeeesong replied:
네트워크 객체에서 에러를 핸들링하도록 바꾸었습니다!

GangWoon commented on IOS/airbnb/airbnb/Network/CustomError.swift:
```diff
@@ -0,0 +1,20 @@
+
+import Foundation
+
+enum CustomError: String, Error, CustomStringConvertible {
```
이름을 갖고 어떤 에러인지 유추하기 가장 어려운 거 같습니다.

eeeesong replied:
넵 적절한 네이밍을 생각해보겠습니다!

GangWoon commented on IOS/airbnb/airbnb/Network/Fake/FakeNetworkManager.swift:
```diff
@@ -0,0 +1,53 @@
+import Foundation
+import Alamofire
+
+final class FakeNetworkManager: AlamofireNetworkManagable {
```
거의 사실상 테스트 코드를 목데이터 처럼 작성하셨습니다 ㅋㅋㅋ
이 부분을 사용해서 테스트 코드를 작성해보시는 것도 좋은 연습이 되실 거 같습니다.
특히 비동기 작업은 관련해서 테스트 코드가 필수적이라고 생각합니다.

eeeesong replied:
그렇군요!!ㅎㅎ 그럼 이 부분을 활용해서 테스트 코드도 작성을 해보도록 하겠습니다 :)

GangWoon reviewed (changes requested):
대화를 통해서 말씀드린 부분과 코맨트로 작성한 부분을 고치시다가 이해가 안가거나 도움이 필요하실 땐 편안하게 질문하셨으면 좋겠습니다.

아 그리고 서치바에서 지우기를 누르지 않고 키보드를 통해서 지울 때 지우기가 비활성화되지 않는 버그가 존재합니다.

eeeesong commented:
미처 발견하지 못한 버그가 있었군요😅 키보드로 지울 때도 버튼이 사라질 수 있도록 수정하도록 하겠습니다.

세심하게 피드백 주셔서 감사합니다. 남겨주신 코멘트 꼼꼼하게 확인해보고, 궁금한 점 있으면 질문 남기도록 하겠습니다.

어제 말씀해주신대로 View Model 부분을 제대로 고치려면 꽤 시간이 필요할 것 같네요! 잘 공부해보고 주말동안 열심히 수정해보겠습니다.

감사합니다!

eeeesong committed:
refactor: ViewModel 역할 개선

- ViewController에서 보관하던 정보, 로직 전부 ViewModel로 이동
- ViewModel 역할에 따라 공용 인터페이스 구축

Dae-Hwa/airbnb/#25

eeeesong committed:
refactor: 불필요 구문 삭제

- 쓰지 않는 메소드, 중복 삭제

Dae-Hwa/airbnb/#25

eeeesong committed:
refactor: 에러 타입 처리 객체 변경

- Network Manager에서 다루도록 변경

Dae-Hwa/airbnb/#25

eeeesong committed:
refactor: 코드 형식 통일

Dae-Hwa/airbnb/#25

eeeesong committed:
refactor: 프로토콜, 메소드 등 리네이밍

Dae-Hwa/airbnb/#25

eeeesong committed:
test: 네트워크 테스트 추가

- PopularLocation 데이터 받아오기 성공한 경우
- 받아오지 못하고 400 오류가 뜬 경우

Dae-Hwa/airbnb/#25

eeeesong committed:
fix: 지우기 버튼 사라지지 않는 버그 수정

- 키보드로 장소 검색 입력란 비울 시, 지우기 버튼이 사라지지 않는 버그 수정

Dae-Hwa/airbnb/#26

eeeesong committed:
Merge pull request #30 from Dae-Hwa/iOS/fix/first-PR

[iOS] 첫번째 PR 수정

eeeesong commented:
@GangWoon 안녕하세요!
주신 피드백을 참고하여 최대한 반영을 해보았습니다.
ViewModel의 역할을 공고화하려고 노력하였고, 협업 차원에서 위험성이 있는 코드들도 수정을 해보았습니다.
(주신 피드백 댓글 아래에 상세 댓글 남겨두었습니다 :))
또한, 거의 처음 작성한 것이라 아직 많이 부족하지만 테스트 코드도 작성을 해보았습니다! 

감사합니다!

GangWoon commented:
ViewModel 변경되면서 책임이 확실하게 변경됐네요 :) 
이제 viewModel에서 중복적으로 사용되는 메소드를 프로토콜 기본 구현으로 작성해주시면 중복코드도 줄일 수 있을겁니다.

그리고 제가 다른 디자인 패턴과 혼동을해서 잘못가이드 드린 부분이 있습니다.
해당 구조에선 뷰를 추상화시킬 수 없는 구조이기때문에 해당 문제를 갖고 고민하셨다면 죄송합니다.

테스트 코드는 내일 제가 코드를 수정해가면서 제가 원하는 방향이 뭔지 예시로 보여드려볼게요.

중간중간 사소한 실수를 제외하면 이미 충분히 코드를 잘 다듬으신 거 같습니다. 
ViewController를 생성하고 dependency를 생성하는 것보다 외부로 부터 주입받는 형태가 좋다는걸 항상 생각하시고 추후에 고치실 여력이 됐을 때 해보시면 좋을 거 같습니다.

리펙토링은 순차적으로 발생하게 됩니다. 이전에 문제점을 고치셨기에 그 다음 문제점을 볼 수 있게 된겁니다. 

해당 피알은 여기서 마무리해도 괜찮을 거 같습니다. 
정말 잘하셨고, 앞으로 남은 기간동안 규모가 커지는 코드를 어떻게 관리할지 같이 고민해보는 시간이 됐으면 좋겠습니다.

eeeesong commented:
리팩토링은 순차적으로 발생한다는 말씀이 굉장히 인상깊네요! 잘 기억하고 있겠습니다 :)

앞으로의 작성에서는 ViewController의 dependency에 대해 더 고민을 해보겠습니다.
그리고 앞서 조언해주신 것처럼 너무 많은 양의 코드를 PR 보내지 않도록 유의해서 다음 PR 때 뵙도록 하겠습니다!

그럼 괜찮으시다면 merge를 부탁드리겠습니다! (저한테는 권한이 없네요😅)
감사합니다!!
